### ISSUE TITLE
Ability to set allowed/denied hosts

### ISSUE THREAD

**Issue by doroved** [SenseUnit/dumbproxy#52]
Please make it possible to set allowed/denied hosts lists for HTTP/HTTPS proxies

- `--allowed-hosts` - list of allowed hosts, to which user can connect via proxy (comma-separated, without spaces, for example - `"google.com,*.google.com,fb.com"`). All other hosts will be denied, if this parameter is provided
- `--denied-hosts` - list of denied hosts (comma-separated, without spaces, for example - `"google.com,*.google.com,fb.com"`). All others hosts will be allowed, if this parameter is provided

**Comment by Snawoot** [comment]
Hello,

I thought about such feature, but it won't do any good. Client can resolve requested host to IP address somehow and request the host by IP address instead of domain name, so it can be trivially circumvented. Overall feature will be unsound, so I won't do it.

**Comment by Snawoot** [comment]
Hi, @doroved 

No, I understood you.

> How are you going to get access to google not by host name, but by its IP?

### Example of normal access of www.google.com

```
curl -vx 127.0.0.1:8080 -o /dev/null https://www.google.com
```

Note that curl sends CONNECT request to www.google.com:443:

```
* Connected to 127.0.0.1 (127.0.0.1) port 8080
* CONNECT tunnel: HTTP/1.1 negotiated
* allocate connect buffer
* Establish HTTP proxy tunnel to www.google.com:443
> CONNECT www.google.com:443 HTTP/1.1
> Host: www.google.com:443
> User-Agent: curl/8.5.0
> Proxy-Connection: Keep-Alive
> 
< HTTP/1.1 200 OK
< 
```

Record in dumbproxy log:

```
PROXY   : 2024/03/03 14:26:47 handler.go:138: INFO     Request: 127.0.0.1:60244 => 127.0.0.1:8080 "" HTTP/1.1 CONNECT //www.google.com:443
```

As you can see hostname is present and can be filtered.

### Example of access to www.google.com circumventing domain filter

Let's resolve the address:

```
[14:29:57] dt1:~> host google.com
google.com has address 142.250.203.206
google.com has IPv6 address 2a00:1450:401b:810::200e
google.com mail is handled by 10 smtp.google.com.
```

Let's make proxy client to connect by IP address, but send all headers and SNI as usual:

```
curl -vx 127.0.0.1:8080 -o /dev/null --connect-to www.google.com:443:142.250.203.206:443 https://www.google.com
```

Now curl sends CONNECT request by IP address:

```
* Connected to 127.0.0.1 (127.0.0.1) port 8080
* CONNECT tunnel: HTTP/1.1 negotiated
* allocate connect buffer
* Establish HTTP proxy tunnel to 142.250.203.206:443
> CONNECT 142.250.203.206:443 HTTP/1.1
> Host: 142.250.203.206:443
> User-Agent: curl/8.5.0
> Proxy-Connection: Keep-Alive
> 
< HTTP/1.1 200 OK
< 
```

Record in dumbproxy log:

```
PROXY   : 2024/03/03 14:30:58 handler.go:138: INFO     Request: 127.0.0.1:37834 => 127.0.0.1:8080 "" HTTP/1.1 CONNECT //142.250.203.206:443
```

As you can see, dumbproxy sees no domain name in there request. And if we remove `-o /dev/null` we will see actual correct response body.

Bottom line is: proposed implementation of this feature is unsound from security perspective.

**Comment by Snawoot** [comment]
> @Snawoot I set up a list of allowed hosts, e.g.
> 
> ```
> google.com
> facebook.com
> ```
> 
> Any request through our proxy if host != google.com|facebook.com -> 400 response

It may work for your use case well, but won't work for proxy clients which do DNS resolving on their own. E.g. Android proxy clients like NekoBox, Adguard, ... make DNS resolving on their own and always send IP address to the proxy. Access logs from such clients look like this:

```
Mar 03 14:11:48 utility dumbproxy[735]: PROXY   : 2024/03/03 14:11:48 handler.go:138: INFO     Request: 109.122.30.107:40706 => 151.115.111.194:443 "mobile" HTTP/2.0 CONNECT //cp.cloudflare.com:80
Mar 03 14:12:25 utility dumbproxy[735]: PROXY   : 2024/03/03 14:12:25 handler.go:138: INFO     Request: 109.122.30.107:52488 => 151.115.111.194:443 "mobile" HTTP/2.0 CONNECT //cp.cloudflare.com:80
Mar 03 14:12:26 utility dumbproxy[735]: PROXY   : 2024/03/03 14:12:26 handler.go:138: INFO     Request: 109.122.30.107:52488 => 151.115.111.194:443 "mobile" HTTP/2.0 CONNECT //8.8.4.4:443
Mar 03 14:12:26 utility dumbproxy[735]: PROXY   : 2024/03/03 14:12:26 handler.go:138: INFO     Request: 109.122.30.107:52488 => 151.115.111.194:443 "mobile" HTTP/2.0 CONNECT //142.250.27.188:5228
Mar 03 14:12:27 utility dumbproxy[735]: PROXY   : 2024/03/03 14:12:27 handler.go:138: INFO     Request: 109.122.30.107:52488 => 151.115.111.194:443 "mobile" HTTP/2.0 CONNECT //69.10.180.70:443
Mar 03 14:12:28 utility dumbproxy[735]: PROXY   : 2024/03/03 14:12:28 handler.go:138: INFO     Request: 109.122.30.107:52488 => 151.115.111.194:443 "mobile" HTTP/2.0 CONNECT //142.250.179.180:443
Mar 03 14:12:31 utility dumbproxy[735]: PROXY   : 2024/03/03 14:12:31 handler.go:138: INFO     Request: 109.122.30.107:52488 => 151.115.111.194:443 "mobile" HTTP/2.0 CONNECT //140.82.121.3:443
Mar 03 14:12:31 utility dumbproxy[735]: PROXY   : 2024/03/03 14:12:31 handler.go:138: INFO     Request: 109.122.30.107:52488 => 151.115.111.194:443 "mobile" HTTP/2.0 CONNECT //185.199.108.154:443
```

Deny list won't work as expected, allow list won't work with all clients at all. Just that alone doesn't make sense to ship such feature.

Another thing is demand for such limitation. Why would one restrict proxy users at all? Can you explain why you need this?

**Comment by Snawoot** [comment]
Well, none of this use cases is a goal for this project. If you're going to provide some commercial services on top of it, you'll likely have to fork it and adapt to your requirements anyway.

If you're concerned about security, then see my remarks above and consider one of following:

* Use certificates authentication instead of password-based auth, dumbproxy supports it. If client does not support it, use stunnel or [steady-tun](https://github.com/Snawoot/steady-tun) as an adapter.
* Consider some form of sandboxing or traffic filtering to restrict the proxy.
* Your own custom fork is an option too.
